### PR TITLE
Automatically switch IME mode between text input and raw input on Windows

### PR DESCRIPTION
**Pull request by Qrox**
#### Summary
SUMMARY: I18N "Automatically switch IME mode between text input and raw input on Windows"

#### Purpose of change
Users of non-alphabetical languages, such as Chinese, Japanese, etc, usually employ an IME to input text. When the user wants to input text in their language, they switch to the composition mode of IME, which intercepts keyboard input and converts it to actual text in the user's language. This leads to problems in CDDA where the users need to constantly switch between composition mode and raw input mode of an IME when inputing text and using hotkeys. It would be more convenient if the game automatically switches the IME mode when starting/ending text input. This PR implements such automatical switching on Windows.

#### Describe the solution
Use the native input method manager (IMM) library of Windows to switch to composition mode when starting text input, such as filter string input, overmap note editing, etc., and automatically restore the previous IME mode when finishing text input.

#### Additional context
Currently several UIs (keybindings menu, world creation menu, character creation menu) are not supported because they accept hotkeys and text input simultaneously, so it's hard to tell whether the user wants to input text or a hotkey. These UIs currently only restores the previous IME mode on return so the user doesn't have to switch back to the previous IME mode when they are finished with the UI.

Additionally, in the current implementation the user can still manually switch to composition mode when the game expects a hotkey. I tried to completely disable composition mode under such situation but unfortunately such functionality does not seem to be supported by the Windows API. It is however desirable since, as mentioned before, some UIs are not automatically switching the IME mode, so the user should be able to manually start text editing in such UIs.

I also need help on testing this PR on Android and Linux to see if I didn't break anything, since I don't have test environments of these platforms.

### CODE REVIEW

File: src/ime.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+#endif
+
+static bool ime_enabled()
+{
+#if defined( __ANDROID__ )
+    return false; // always call disable_ime() (i.e. do nothing) on return
+#elif defined( _WIN32 )
+    return imm.ime_enabled();
+#endif
+    // TODO: other platforms?
```
Comment (BevapDin) [inline src/ime.cpp]: At the very least add a `return false;` here so the compiler won't complain abount missing return value.

File: src/sdltiles.cpp [hunk 2]
```diff
@@ -24,8 +24,10 @@
 
 #if defined(_MSC_VER) && defined(USE_VCPKG)
 #   include <SDL2/SDL_image.h>
+#   include <SDL2/SDL_syswm.h>
 #else
 #   include <SDL_image.h>
+#   include <SDL_syswm.h>
```
Comment (nshcat) [inline src/sdltiles.cpp]: This include messes with Linux builds -- it pulls in X11's `X.h`, which causes a multiple definition error of `class Font`.

Comment (nshcat) [reply src/sdltiles.cpp]:
```
src/sdltiles.cpp:104:7: error: definition of type 'Font' conflicts with typedef of the same name
class Font
      ^
/usr/include/X11/X.h:100:13: note: 'Font' declared here
typedef XID Font;
            ^
```

Comment (Qrox) [reply src/sdltiles.cpp]: Thanks, fixed!

**Comment by nexusmrsep** [comment]
image
Can't compile master after this change.

**Comment by Qrox** [comment]
Curious, it passed all CI builds without problem (apart from gorgon, as it failed for irrelevant reasons). What happens if you remove `WINAPI` from the typedefs?

**Comment by ZhilkinSerg** [comment]
He's using native MinGW/C::B build environment which is not covered by CI (yet).

**Comment by Qrox** [comment]
After some search it seems that the problem is that HIMC isn't defined anywhere. @nexusmrsep Could you try to add `#include <imm.h>` and/or `#include <Immdev.h>` to the start of `ime.cpp` and see if it compiles?

**Comment by ZhilkinSerg** [comment]
```C++
#include "Windows.h"
#include <imm.h>
```

in `ime.h` should do the trick probably

**Comment by ZhilkinSerg** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/pull/33227 should compile under C::B and MinGW.

**Pull request merged**